src/rt/cpp/when.h

```diff
@@ -100,25 +94,44 @@ namespace verona::cpp
           sizeof...(Args),
           cowns,
           [f = std::forward<F>(f), cown_tuple = cown_tuple]() mutable {
-            /// Effectively converts cown_ptr... to acquired_cown... .
-            auto lift_f = [f = std::forward<F>(f)](Args... args) mutable {
-              f(cown_ptr_to_acquired(args)...);
-            };
+            /// Effectively converts ActualCown<T>... to
+            /// acquired_cown... .
+            auto lift_f =
+              [f = std::forward<F>(f)](ActualCown<Args>*... args) mutable {
+                f(actual_cown_to_acquired<Args>(args)...);
+              };
 
             std::apply(lift_f, cown_tuple);
           });
       }
     }
   };
 
+  /**
+   * Internal: required for template inference.

```

Comment:
Do we require C++17?  If so, you can probably do this with a [deduction guide](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) instead of a wrapper function.